Userland/find.cpp

```diff
@@ -0,0 +1,468 @@
+};
+
+class GroupCommand final : public StatCommand {
+public:
+    GroupCommand(const char* arg)
+    {
+        if (struct group* gr = getgrnam(arg)) {
+            m_gid = gr->gr_gid;
+        } else {
+            // Attempt to parse it as decimal GID.
+            bool ok;
+            m_gid = StringView(arg).to_int(ok);
+            if (!ok)
+                fatal_error("Invalid group: \033[1m%s", arg);
+        }
+    }
+
+private:
+    virtual bool evaluate(const struct stat& stat) const override
+    {
+        return stat.st_gid == m_gid;
+    }
+
+    gid_t m_gid { 0 };
+};
+
+class SizeCommand final : public StatCommand {
+public:
+    SizeCommand(const char* arg)
+    {
+        StringView view = arg;
+        if (view.ends_with('c')) {
+            m_is_bytes = true;
+            view = view.substring_view(0, view.length() - 1);
+        }
+        bool ok;
+        m_size = view.to_uint(ok);
+        if (!ok)
+            fatal_error("Invalid size: \033[1m%s", arg);
+    }
+
+private:
+    virtual bool evaluate(const struct stat& stat) const override
+    {
+        if (m_is_bytes)
+            return stat.st_size == m_size;
+
+        auto size_divided_by_512_rounded_up = (stat.st_size + 511) / 512;
+        return size_divided_by_512_rounded_up == m_size;
+    }
+
+    off_t m_size { 0 };
+    bool m_is_bytes { false };
+};
+
+class PrintCommand final : public Command {
+public:
+    PrintCommand(char terminator = '\n')
+        : m_terminator(terminator)
+    {
+    }
+
+private:
+    virtual bool evaluate(const char* file_path) const override
+    {
+        printf("%s%c", file_path, m_terminator);
+        return true;
+    }
+
+    char m_terminator { '\n' };
+};
+
+class ExecCommand final : public Command {
+public:
+    ExecCommand(Vector<char*>&& argv)
+        : m_argv(move(argv))
+    {
+    }
+
+private:
+    virtual bool evaluate(const char* file_path) const override
+    {
+        pid_t pid = fork();
+
+        if (pid < 0) {
+            perror("fork");
+            g_there_was_an_error = true;
+            return false;
+        } else if (pid == 0) {
+            // Replace any occurrences of "{}" with the path. Since we're in the
+            // child and going to exec real soon, let's just const_cast away the
+            // constness.
+            auto argv = const_cast<Vector<char*>&>(m_argv);
+            for (auto& arg : argv) {
+                if (StringView(arg) == "{}")
+                    arg = const_cast<char*>(file_path);
+            }
+            argv.append(nullptr);
+            execvp(m_argv[0], argv.data());
+            perror("execvp");
+            exit(1);
+        } else {
+            int status;
+            int rc = waitpid(pid, &status, 0);
+            if (rc < 0) {
+                perror("waitpid");
+                g_there_was_an_error = true;
+                return false;
+            }
+            return WIFEXITED(status) && WEXITSTATUS(status) == 0;
+        }
```

Comment:
What if it was signaled? Are we gonna leave it to run in the background?

Reply:
wdym, if it was killed by a signal, we'll reap it (and evaluate to false), otherwise we'll keep waiting for it.

Reply:
Ah, nvm, I got it mixed up.